### PR TITLE
Introduce SafePyObject, make TorchDispatchTypeObject use it

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #75161
* #75154
* #75159
* __->__ #75142
* #75141

The pattern of a PyObject* bundled with a PyInterpreter* is pretty
useful in many contexts (e.g., TorchDispatchTypeObject) so I have turned
it into a dedicated class SafePyObject.  In the process I fixed a
bug with the old TorchDispatchTypeObject (copy constructor/assignment
was not deleted), made the API more safe (retrieving the PyObject*
pointer requires verification that the PyInterpreter* matches) and
fixed some minor inefficiencies in C++ code.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

### CODE REVIEW

**Comment by ezyang** [comment]
@anjali411 this might help for your stuff, this is the thing you discussed with @bdhirsh and @albanD

File: c10/core/SafePyObject.h [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+//    you do not have a libpython dependency
+//
+//  - It is multi-interpreter safe (ala torchdeploy); when you fetch
+//    the underlying PyObject* you are required to specify what the current
+//    interpreter context is and we will check that you match it.
+//
+// It is INVALID to store a reference to a Tensor object in this way;
+// you should just use TensorImpl directly in that case!
+struct C10_API SafePyObject {
+  // Steals a reference to type_object
```
Comment (zou3519) [inline c10/core/SafePyObject.h:22]: do you mean `data`?

**Review by zou3519 — APPROVED** [review]
cool!

File: c10/core/SafePyObject.h (hunk 1 above)
Comment (albanD) [reply c10/core/SafePyObject.h:22]: Is the steal semantic the same as `py::object`? From https://github.com/pybind/pybind11/issues/1201, it doesn't look like it? Should that be a third major difference listed above?

**Review by albanD — APPROVED** [review]
cc @anjali411 this would most likely help you!

Change looks good, maybe a small doc update.

File: c10/core/SafePyObject.h [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+  SafePyObject(PyObject* data, c10::impl::PyInterpreter* pyinterpreter)
+      : data_(data), pyinterpreter_(pyinterpreter) {}
+
+  // In principle this could be copyable if we add an incref to PyInterpreter
+  // but for now it's easier to just disallow it.
+  SafePyObject(SafePyObject const&) = delete;
```
[the remark this answers is not in the document]
Comment (ezyang) [reply c10/core/SafePyObject.h:28]: if you want to have multiple references to the same PyObject. This situation happens in mode use, although we put the SafePyObject in a shared ptr so we can conveniently share it in C++ without having to take out the GIL

File: c10/core/SafePyObject.h (hunk 1 above)
Comment (ezyang) [reply c10/core/SafePyObject.h:22]: py::object just doesn't accept a bare PyObject pointer at all. I suppose if you wanted to be safe we could add explicit borrow/steal APIs...

**Pull request closed without merging**